singularity-eos/eos/eos_eospac.hpp

```diff
@@ -72,103 +72,1156 @@ class EOSPAC : public EosBase<EOSPAC> {
+
+    for (int i = 0; i < num; ++i) {
+      R[i] = rhos[i];
+      E[i] = sieToSesame(sies[i]);
+    }
+
+    EOS_INTEGER table = TofRE_table_;
+    eosSafeInterpolate(&table, num, R, E, T, dTdr, dTde, "TofRE", Verbosity::Quiet);
+
+    for (int i = 0; i < num; ++i) {
+      temperatures[i] = temperatureFromSesame(T[i]);
+    }
+  }
+
+  template <int CHUNK_SIZE = 64, typename RealIndexer, typename ConstRealIndexer,
+            typename LambdaIndexer>
+  inline void PressureFromDensityTemperature(ConstRealIndexer &&rhos,
+                                             ConstRealIndexer &&temperatures,
+                                             RealIndexer &&pressures, const int num,
+                                             LambdaIndexer &&lambdas) const {
+    using namespace EospacWrapper;
+    int nchunks = (num / CHUNK_SIZE) + (num % CHUNK_SIZE ? 1 : 0);
+    std::array<EOS_REAL, CHUNK_SIZE> R, T, P, dPdr, dPdT;
+
+    for (int ichunk = 0; ichunk < nchunks; ++ichunk) {
+      int istart = ichunk * CHUNK_SIZE;
+      int iend = std::min(num, istart + CHUNK_SIZE);
+      int nelements = iend - istart;
+
+      for (int i = 0; i < nelements; ++i) {
+        int j = istart + i;
+        R[i] = rhos[j];
+        T[i] = temperatureToSesame(temperatures[j]);
+      }
+
+      EOS_INTEGER table = PofRT_table_;
+      eosSafeInterpolate(&table, nelements, R.data(), T.data(), P.data(), dPdr.data(),
+                         dPdT.data(), "PofRT", Verbosity::Quiet);
+
+      for (int i = 0; i < nelements; ++i) {
+        int j = istart + i;
+        pressures[j] = pressureFromSesame(P[i]);
+      }
+    }
+  }
+
+  template <typename RealIndexer, typename ConstRealIndexer, typename LambdaIndexer>
+  inline void
+  PressureFromDensityTemperature(ConstRealIndexer &&rhos, ConstRealIndexer &&temperatures,
+                                 RealIndexer &&pressures, Real *scratch, const int num,
+                                 LambdaIndexer &&lambdas) const {
+    using namespace EospacWrapper;
+    EOS_REAL *R = scratch;
+    EOS_REAL *T = scratch + num;
+    EOS_REAL *P = scratch + 2 * num;
+    EOS_REAL *dPdr = scratch + 3 * num;
+    EOS_REAL *dPdT = scratch + 4 * num;
+
+    for (int i = 0; i < num; ++i) {
+      R[i] = rhos[i];
+      T[i] = temperatureToSesame(temperatures[i]);
+    }
+
+    EOS_INTEGER table = PofRT_table_;
+    eosSafeInterpolate(&table, num, R, T, P, dPdr, dPdT, "PofRT", Verbosity::Quiet);
+
+    for (int i = 0; i < num; ++i) {
+      pressures[i] = pressureFromSesame(P[i]);
+    }
+  }
+
+  template <int CHUNK_SIZE = 64, typename RealIndexer, typename LambdaIndexer>
+  inline void FillEos(RealIndexer &&rhos, RealIndexer &&temps, RealIndexer &&energies,
+                      RealIndexer &&presses, RealIndexer &&cvs, RealIndexer &&bmods,
+                      const int num, const unsigned long output,
+                      LambdaIndexer &&lambdas) const {
+    using namespace EospacWrapper;
```

Comment:
It's pretty painful how much duplicated code lives in `FillEos`. I wonder if there's a way to reduce this.

Reply:
Could the chunk version call the regular version but with chunked data?